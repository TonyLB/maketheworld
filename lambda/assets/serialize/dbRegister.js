import { replaceItem, mergeIntoDataRange } from '../utilities/dynamoDB/index.js'

export const dbRegister = async (dbClient, { fileName, translateFile, importTree, scopeMap, assets }) => {
    const asset = assets.find(({ tag }) => (tag === 'Asset'))
    if (asset && asset.key) {
        await Promise.all([
            replaceItem(dbClient, {
                AssetId: `ASSET#${asset.key}`,
                DataCategory: 'Meta::Asset',
                fileName,
                translateFile,
                importTree,
                name: asset.name,
                description: asset.description,
                player: asset.player
            }),
            mergeIntoDataRange({
                dbClient,
                table: 'assets',
                search: { DataCategory: `ASSET#${asset.key}` },
                items: assets
                    .filter(({ tag }) => (tag === 'Room'))
                    .map(({ tag, ...rest }) => (rest)),
                mergeFunction: ({ current, incoming }) => {
                    if (!incoming) {
                        return 'delete'
                    }
                    if (!current) {
                        const { key, isGlobal, ...rest } = incoming
                        return {
                            scopedId: key,
                            ...rest
                        }
                    }
                    //
                    // TODO: When Room entries are expanded to store more than the sheer fact of their
                    // existence (likely as part of Map storage), extend this function to compensate
                    // by testing whether an update is needed
                    //
                    return 'ignore'
                },
                extractKey: (item) => {
                    if (item.isGlobal) {
                        return `ROOM#${item.key}`
                    }
                    if (scopeMap[item.key]) {
                        return `ROOM#${scopeMap[item.key]}`
                    }
                    console.log(`ERROR:  ScopeMap in dbRegister has no entry for ${item.key}`)
                    return `ROOM#${uuidv4()}`
                }
            })
        ])
    }
    const character = assets.find(({ tag }) => (tag === 'Character'))
    if (character && character.key) {
        await Promise.all([
            replaceItem(dbClient, {
                AssetId: `CHARACTER#${scopeMap[character.key]}`,
                DataCategory: 'Meta::Character',
                fileName,
                translateFile,
                scopedId: character.key,
                ...(['Name', 'Pronouns', 'FirstImpression', 'OneCoolThing', 'Outfit', 'player']
                    .reduce((previous, label) => ({ ...previous, [label]: character[label] }), {}))
            })
        ])
    }

}
