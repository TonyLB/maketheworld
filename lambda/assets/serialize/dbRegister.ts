import { v4 as uuidv4 } from 'uuid'
import { assetDB } from '@tonylb/mtw-utilities/dist/dynamoDB/index.js'
import { AssetKey } from '@tonylb/mtw-utilities/dist/types.js'
import AssetWorkspace from '@tonylb/mtw-asset-workspace/dist/index.js'
import { isNormalAsset, isNormalComponent, isNormalExit, NormalForm, isNormalCharacter, NormalItem, isNormalMap, isNormalRoom, isNormalFeature, NormalReference, isNormalImport } from '@tonylb/mtw-wml/dist/normalize/baseClasses.js'
import { EphemeraCharacterId, EphemeraError } from '@tonylb/mtw-interfaces/dist/baseClasses.js'
import internalCache from '../internalCache'
import messageBus from '../messageBus'
import setEdges from '@tonylb/mtw-utilities/dist/graphStorage/update/setEdges'
import { graphDBHandler } from '../internalCache/graph'
import { graphStorageDB } from './graphCache'
import { CharacterKey } from '@tonylb/mtw-utilities/dist/types'

const tagRenderLink = (normalForm) => (renderItem) => {
    if (typeof renderItem === 'object') {
        if (renderItem.tag === 'Link') {
            return {
                ...renderItem,
                targetTag: normalForm[renderItem.to]?.tag
            }
        }
    }
    return renderItem
}

const denormalizeExits = (normalForm: NormalForm) => (contents: NormalReference[]) => {
    const exitTags = contents.filter(({ tag }) => (tag === 'Exit'))
    const exits = exitTags.map(({ key }) => (normalForm[key])).filter(isNormalExit)
    return exits.map(({ name, to }) => ({ name, to }))
}

const noConditionContext = ({ contextStack }) => (!contextStack.find(({ tag }) => (tag === 'If')))

const itemRegistry = (normalForm: NormalForm) => (item: NormalItem) => {
    if (isNormalMap(item)) {
        const mapDefaultAppearances = item.appearances
            .filter(noConditionContext)
            .map(({ rooms }) => ({
                rooms: rooms.map(({ key, ...rest }) => {
                    const lookup = normalForm[key]
                    if (!isNormalRoom(lookup)) {
                        throw new EphemeraError(`Invalid item in map room lookup: ${key}`)
                    }
                    return {
                        ...rest,
                        name: (lookup.appearances || [])
                            .filter(noConditionContext)
                            .map<string>(({ name }) => ((name || []).map((namePart) => (namePart.tag === 'String' ? namePart.value : '')).join('')))
                            .join('')
                    }
                })
            }))
        return {
            tag: item.tag,
            key: item.key,
            defaultAppearances: mapDefaultAppearances
        }
    }
    if (isNormalRoom(item)) {
        return {
            tag: item.tag,
            key: item.key,
            defaultAppearances: item.appearances
                .filter(noConditionContext)
                .filter(({ contents= [], render = [], name = [] }) => (contents.length > 0 || render.length > 0 || name.length > 0))
                .map(({ contents = [], render = [], name = [] }) => ({
                    exits: denormalizeExits(normalForm)(contents),
                    render: render.map(tagRenderLink(normalForm)),
                    name: name.map((item) => (item.tag === 'String' ? item.value : '')).join('')
                }))
        }
    }
    if (isNormalFeature(item)) {
        return {
            tag: item.tag,
            key: item.key,
            defaultAppearances: item.appearances
                .filter(noConditionContext)
                .filter(({ contents= [], render = [], name = [] }) => (contents.length > 0 || render.length > 0 || name.length > 0))
                .map(({ render, name = [] }) => ({
                    render: (render || []).map(tagRenderLink(normalForm)),
                    name: name.map((item) => (item.tag === 'String' ? item.value : '')).join('')
                }))
        }
    }
}

export const dbRegister = async (assetWorkspace: AssetWorkspace): Promise<void> => {
    const { address } = assetWorkspace
    // if (assetWorkspace.status.json !== 'Clean') {
    //     await assetWorkspace.loadJSON()
    // }
    const assets: NormalForm = assetWorkspace.normal || {}
    const asset = Object.values(assets).find(isNormalAsset)
    if (asset && asset.key) {
        const defaultExits = Object.values(assets)
            .filter(isNormalExit)
            .filter(({ appearances }) => ((appearances || []).find(noConditionContext)))
            .map(({ name, to, from }) => ({ name, to, from }))
        const defaultNames = Object.values(assets)
            .filter(isNormalComponent)
            .map(({ tag, key, appearances }) => ({
                tag,
                key,
                name: (appearances || [])
                    .filter(noConditionContext)
                    .map(({ name = [] }) => (name.map((item) => (item.tag === 'String' ? item.value : '')).join('')))
                    .join('')
            }))
            .filter(({ name }) => (Boolean(name)))
            .map(({ tag, key, name }) => ({ [key]: { tag, name } }))
            .reduce((previous, entry) => (Object.assign(previous, entry)), {})
        const updatedLibraryAssets = {
            [AssetKey(asset.key)]: {
                AssetId: AssetKey(asset.key),
                scopedId: asset.key,
                Story: asset.Story,
                instance: asset.instance,
            }
        }
        const updatedPlayerAssets = {
            [asset.key]: {
                AssetId: asset.key,
                scopedId: asset.key,
                Story: asset.Story,
                instance: asset.instance,
            }
        }
        const updateLibraryPromise = address.zone === 'Personal'
            ? internalCache.PlayerLibrary.set(address.player, {
                Assets: updatedPlayerAssets,
                Characters: {}
            })
            : address.zone === 'Library'
                ? internalCache.Library.set({
                    Assets: updatedLibraryAssets,
                    Characters: {}
                })
                : Promise.resolve({})
        await Promise.all([
            setEdges({ internalCache: internalCache._graphCache, dbHandler: graphStorageDB })(
                AssetKey(asset.key),
                Object.values(assets)
                    .filter(isNormalImport)
                    .map(({ from }) => ({ target: AssetKey(from), context: '' })),
                'back'
            ),
            assetDB.putItem({
                AssetId: AssetKey(asset.key),
                DataCategory: `Meta::Asset`,
                address,
                Story: asset.Story,
                instance: asset.instance,
                importTree: [],
                defaultExits,
                defaultNames,
                zone: address.zone,
                ...(address.zone === 'Personal' ? { player: address.player } : {})
            }),
            assetDB.merge({
                query: {
                    IndexName: 'DataCategoryIndex',
                    Key: {
                        DataCategory: AssetKey(asset.key)
                    }
                },
                items: asset.instance
                    ? []
                    : Object.values(assets)
                        .filter(({ tag }) => (['Room', 'Feature', 'Map'].includes(tag)))
                        .map(itemRegistry(assets)) as any,
                mergeFunction: ({ current, incoming }) => {
                    if (!incoming) {
                        return 'delete'
                    }
                    if (!current || (JSON.stringify(current) !== JSON.stringify(incoming))) {
                        const { tag, key, ...rest } = incoming
                        return {
                            scopedId: key,
                            ...rest
                        }
                    }

                    return 'ignore'
                },
                extractKey: ({ tag, key }) => {
                    let prefix = ''
                    switch(tag) {
                        case 'Feature':
                        case 'Map':
                        case 'Bookmark':
                        case 'Message':
                        case 'Moment':
                            prefix = tag.toUpperCase()
                            break
                        default:
                            prefix = 'ROOM'
                    }
                    if (assetWorkspace._isGlobal) {
                        return `${prefix}#${key}`
                    }
                    if (assetWorkspace.namespaceIdToDB[key]) {
                        return assetWorkspace.namespaceIdToDB[key]
                    }
                    console.log(`ERROR:  ScopeMap in dbRegister has no entry for ${key}`)
                    return `${prefix}#${uuidv4()}`
                }
            }),
            updateLibraryPromise
        ])
    }
    const character = Object.values(assets).find(isNormalCharacter)
    if (character && character.key) {
        const images = (character.images || [])
            .map((image) => (assetWorkspace.properties[image]?.fileName))
            .filter((image) => (image))
        const updatedCharacters = {
            [assetWorkspace.namespaceIdToDB[character.key]]: {
                CharacterId: assetWorkspace.namespaceIdToDB[character.key] as EphemeraCharacterId,
                Name: character.Name,
                FirstImpression: character.FirstImpression,
                OneCoolThing: character.OneCoolThing,
                Pronouns: character.Pronouns,
                Outfit: character.Outfit,
                fileName: '',
                fileURL: images[0] || undefined,
                scopedId: character.key,        
            }
        }
        const updateLibraryPromise = address.zone === 'Personal'
            ? internalCache.PlayerLibrary.set(address.player, {
                Assets: {},
                Characters: updatedCharacters
            })
            : address.zone === 'Library'
                ? internalCache.Library.set({
                    Assets: {},
                    Characters: updatedCharacters
                })
                : Promise.resolve({})
        await Promise.all([
            setEdges({ internalCache: internalCache._graphCache, dbHandler: graphStorageDB })(
                CharacterKey(character.key),
                Object.values(assets)
                    .filter(isNormalImport)
                    .map(({ from }) => ({ target: AssetKey(from), context: '' })),
                'back'
            ),
            assetDB.putItem({
                AssetId: assetWorkspace.namespaceIdToDB[character.key],
                DataCategory: `Meta::Character`,
                address,
                zone: address.zone,
                Name: character.Name,
                FirstImpression: character.FirstImpression,
                OneCoolThing: character.OneCoolThing,
                Pronouns: character.Pronouns,
                Outfit: character.Outfit,
                images,
                scopedId: character.key,
                ...(address.zone === 'Personal' ? { player: address.player } : {})
            }),
            updateLibraryPromise
        ])
        if (address.zone === 'Personal') {
            messageBus.send({
                type: 'PlayerInfo',
                player: address.player
            })

        }
        if (address.zone === 'Library') {
            messageBus.send({ type: 'LibraryUpdate' })
        }
    }

}
