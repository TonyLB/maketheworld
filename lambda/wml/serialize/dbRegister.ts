import { assetDB } from '@tonylb/mtw-utilities/dist/dynamoDB/index.js'
import { AssetKey } from '@tonylb/mtw-utilities/dist/types.js'
import ReadOnlyAssetWorkspace from '@tonylb/mtw-asset-workspace/dist/readOnly'
import { isNormalAsset, NormalForm, isNormalCharacter, isNormalImport } from '@tonylb/mtw-wml/dist/normalize/baseClasses'
import { graphCache, graphStorageDB } from './graphCache'
import { CharacterKey } from '@tonylb/mtw-utilities/dist/types'
import GraphUpdate from '@tonylb/mtw-utilities/dist/graphStorage/update'
import { snsClient } from '../clients'
import { PublishCommand } from '@aws-sdk/client-sns'

const { FEEDBACK_TOPIC } = process.env

export const dbRegister = async (assetWorkspace: ReadOnlyAssetWorkspace): Promise<void> => {
    const { address } = assetWorkspace
    const assets: NormalForm = assetWorkspace.normal || {}
    const asset = Object.values(assets).find(isNormalAsset)
    if (asset && asset.key) {
        const graphUpdate = new GraphUpdate({ internalCache: graphCache, dbHandler: graphStorageDB })
        graphUpdate.setEdges([{
            itemId: AssetKey(asset.key),
            edges: Object.values(assets)
                .filter(isNormalImport)
                .map(({ from }) => ({ target: AssetKey(from), context: '' })),
            options: { direction: 'back' }
        }])
        await Promise.all([
            graphUpdate.flush(),
            assetDB.putItem({
                AssetId: AssetKey(asset.key),
                DataCategory: `Meta::Asset`,
                address,
                Story: asset.Story,
                instance: asset.instance,
                zone: address.zone,
                ...(address.zone === 'Personal' ? { player: address.player } : {})
            })
        ])
    }
    const character = Object.values(assets).find(isNormalCharacter)
    if (character && character.key) {
        const images = (character.images || [])
            .map((image) => (assetWorkspace.properties[image]?.fileName))
            .filter((image) => (image))
        const graphUpdate = new GraphUpdate({ internalCache: graphCache, dbHandler: graphStorageDB })
        graphUpdate.setEdges([{
            itemId: CharacterKey(character.key),
            edges: Object.values(assets)
                .filter(isNormalImport)
                .map(({ from }) => ({ target: AssetKey(from), context: '' })),
            options: { direction: 'back' }
        }])
        await Promise.all([
            graphUpdate.flush(),
            assetDB.putItem({
                AssetId: assetWorkspace.namespaceIdToDB[character.key],
                DataCategory: `Meta::Character`,
                address,
                zone: address.zone,
                Name: character.Name,
                FirstImpression: character.FirstImpression,
                OneCoolThing: character.OneCoolThing,
                Pronouns: character.Pronouns,
                Outfit: character.Outfit,
                images,
                scopedId: character.key,
                ...(address.zone === 'Personal' ? { player: address.player } : {})
            })
        ])
        if (address.zone === 'Personal') {
            await snsClient.send(new PublishCommand({
                TopicArn: FEEDBACK_TOPIC,
                Message: JSON.stringify({ player: address.player }),
                MessageAttributes: {
                    Type: { DataType: 'String', StringValue: 'PlayerInfo' }
                }
            }))
        }
        if (address.zone === 'Library') {
            await snsClient.send(new PublishCommand({
                TopicArn: FEEDBACK_TOPIC,
                Message: '{}',
                MessageAttributes: {
                    Type: { DataType: 'String', StringValue: 'LibraryUpdate' }
                }
            }))
        }
    }

}
