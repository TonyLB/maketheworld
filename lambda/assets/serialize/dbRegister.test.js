import { jest, describe, it, expect } from '@jest/globals'

import { assetDB, mergeIntoDataRange } from '/opt/utilities/dynamoDB/index.js'

import { dbRegister } from './dbRegister.js'

describe('dbRegister', () => {
    beforeEach(() => {
        jest.clearAllMocks()
        jest.resetAllMocks()
    })
    it('should put a single element for a Character file', async () => {
        await dbRegister({
            fileName: 'test.wml',
            translateFile: 'test.translate.json',
            scopeMap: { TESS: '12345' },
            assets: {
                TESS: {
                    tag: 'Character',
                    key: 'TESS',
                    zone: 'Library',
                    Name: 'Tess',
                    fileURL: 'testIcon.png',
                    Pronouns: {
                        subject: 'she',
                        object: 'her',
                        possessive: 'her',
                        adjective: 'hers',
                        reflexive: 'herself'
                    },
                    FirstImpression: 'Frumpy Goth',
                    OneCoolThing: 'Fuchsia eyes',
                    Outfit: 'A bulky frock-coat lovingly kit-bashed from a black hoodie and patchily dyed lace.',
                    player: 'TEST'
                }
            }
        })
        expect(assetDB.putItem).toHaveBeenCalledWith({
            AssetId: 'CHARACTER#12345',
            DataCategory: 'Meta::Character',
            zone: 'Library',
            fileName: 'test.wml',
            translateFile: 'test.translate.json',
            scopedId: 'TESS',
            player: 'TEST',
            fileURL: 'testIcon.png',
            Name: 'Tess',
            Pronouns: {
                subject: 'she',
                object: 'her',
                possessive: 'her',
                adjective: 'hers',
                reflexive: 'herself'
            },
            FirstImpression: 'Frumpy Goth',
            OneCoolThing: 'Fuchsia eyes',
            Outfit: 'A bulky frock-coat lovingly kit-bashed from a black hoodie and patchily dyed lace.'
        })
    })

    it('should save meta, rooms for Asset type', async () => {
        await dbRegister({
            fileName: 'test.wml',
            translateFile: 'test.translate.json',
            scopeMap: {
                Welcome: '12345',
                power: 'ABC',
                togglePower: 'DEF'
            },
            assets: {
                TEST: {
                    tag: 'Asset',
                    key: 'TEST',
                    name: 'Test',
                    zone: 'Library',
                    appearances: [{
                        contextStack: [],
                        contents: [{
                            tag: 'Map',
                            key: 'Village',
                            index: 0
                        },
                        {
                            tag: 'Room',
                            key: 'Welcome',
                            index: 0
                        },
                        {
                            tag: 'Feature',
                            key: 'clockTower',
                            index: 0
                        },
                        {
                            tag: 'Variable',
                            key: 'power',
                            index: 0
                        },
                        {
                            tag: 'Action',
                            key: 'togglePower',
                            index: 0
                        }]
                    }]
                },
                Village: {
                    tag: 'Map',
                    key: 'Village',
                    appearances: [{
                        contextStack: [{ tag: 'Asset', key: 'TEST', index: 0 }],
                        contents: []
                    }]
                },
                Welcome: {
                    tag: 'Room',
                    key: 'Welcome',
                    appearances: [{
                        contextStack: [{ tag: 'Asset', key: 'TEST', index: 0 }],
                        contents: []
                    }]
                },
                clockTower: {
                    tag: 'Feature',
                    key: 'clockTower',
                    appearances: [{
                        contextStack: [{ tag: 'Asset', key: 'TEST', index: 0 }],
                        contents: []
                    }]
                },
                power: {
                    tag: 'Variable',
                    key: 'power',
                    default: true,
                    appearances: [{
                        contextStack: [{ tag: 'Asset', key: 'TEST', index: 0 }],
                        contents: []
                    }]
                },
                togglePower: {
                    tag: 'Action',
                    key: 'togglePower',
                    src: 'power = !power',
                    appearances: [{
                        contextStack: [{ tag: 'Asset', key: 'TEST', index: 0 }],
                        contents: []
                    }]
                }
            }
        })
        expect(assetDB.putItem).toHaveBeenCalledWith({
            AssetId: 'ASSET#TEST',
            DataCategory: 'Meta::Asset',
            fileName: 'test.wml',
            translateFile: 'test.translate.json',
            name: 'Test',
            zone: 'Library'
        })
        expect(mergeIntoDataRange).toHaveBeenCalledWith({
            table: 'assets',
            search: { DataCategory: 'ASSET#TEST' },
            items: [{
                tag: 'Map',
                key: 'Village'
            },
            {
                tag: 'Room',
                key: 'Welcome',
                defaultAppearances: []
            },
            {
                tag: 'Feature',
                key: 'clockTower',
                defaultAppearances: []
            }],
            mergeFunction: expect.any(Function),
            extractKey: expect.any(Function)
        })
    })

    it('should save meta, rooms for Story type', async () => {
        await dbRegister({
            fileName: 'test.wml',
            translateFile: 'test.translate.json',
            scopeMap: {
                Welcome: '12345',
                power: 'ABC',
                togglePower: 'DEF'
            },
            assets: {
                TEST: {
                    tag: 'Asset',
                    Story: true,
                    key: 'TEST',
                    name: 'Test',
                    zone: 'Library',
                    appearances: [{
                        contextStack: [],
                        contents: [{
                            tag: 'Room',
                            key: 'Welcome',
                            index: 0
                        },
                        {
                            tag: 'Variable',
                            key: 'power',
                            index: 0
                        },
                        {
                            tag: 'Action',
                            key: 'togglePower',
                            index: 0
                        }]
                    }]
                },
                Welcome: {
                    tag: 'Room',
                    key: 'Welcome',
                    appearances: [{
                        contextStack: [{ key: 'TEST', tag: 'Asset', index: 0 }],
                        contents: [],
                    }]
                },
                power: {
                    tag: 'Variable',
                    key: 'power',
                    default: true,
                    appearances: [{
                        contextStack: [{ key: 'TEST', tag: 'Asset', index: 0 }],
                        contents: [],
                    }]
                },
                togglePower: {
                    tag: 'Action',
                    key: 'togglePower',
                    src: 'power = !power',
                    appearances: [{
                        contextStack: [{ key: 'TEST', tag: 'Asset', index: 0 }],
                        contents: [],
                    }]
                }
            }
        })
        expect(assetDB.putItem).toHaveBeenCalledWith({
            AssetId: 'ASSET#TEST',
            DataCategory: 'Meta::Asset',
            Story: true,
            fileName: 'test.wml',
            translateFile: 'test.translate.json',
            name: 'Test',
            zone: 'Library'
        })
        expect(mergeIntoDataRange).toHaveBeenCalledWith({
            table: 'assets',
            search: { DataCategory: 'ASSET#TEST' },
            items: [{
                tag: 'Room',
                key: 'Welcome',
                defaultAppearances: []
            }],
            mergeFunction: expect.any(Function),
            extractKey: expect.any(Function)
        })
    })

    it('should save meta only for instanced Story type', async () => {
        await dbRegister({
            fileName: 'test.wml',
            translateFile: 'test.translate.json',
            scopeMap: {
                Welcome: '12345',
                power: 'ABC',
                togglePower: 'DEF'
            },
            assets: {
                TEST: {
                    tag: 'Asset',
                    Story: true,
                    instance: true,
                    key: 'TEST',
                    name: 'Test',
                    zone: 'Library',
                    appearances: [{
                        contextStack: [],
                        contents: [{
                            tag: 'Room',
                            key: 'Welcome',
                            index: 0
                        },
                        {
                            tag: 'Variable',
                            key: 'power',
                            index: 0
                        },
                        {
                            tag: 'Action',
                            key: 'togglePower',
                            index: 0
                        }]
                    }]
                },
                Welcome: {
                    tag: 'Room',
                    key: 'Welcome',
                    appearances: [{
                        contextStack: [{ key: 'TEST', tag: 'Asset', index: 0 }],
                        contents: [],
                    }]
                },
                power: {
                    tag: 'Variable',
                    key: 'power',
                    default: true,
                    appearances: [{
                        contextStack: [{ key: 'TEST', tag: 'Asset', index: 0 }],
                        contents: [],
                    }]
                },
                togglePower: {
                    tag: 'Action',
                    key: 'togglePower',
                    src: 'power = !power',
                    appearances: [{
                        contextStack: [{ key: 'TEST', tag: 'Asset', index: 0 }],
                        contents: [],
                    }]
                }
            }
        })
        expect(assetDB.putItem).toHaveBeenCalledWith({
            AssetId: 'ASSET#TEST',
            DataCategory: 'Meta::Asset',
            Story: true,
            instance: true,
            fileName: 'test.wml',
            translateFile: 'test.translate.json',
            name: 'Test',
            zone: 'Library'
        })
        expect(mergeIntoDataRange).toHaveBeenCalledWith({
            table: 'assets',
            search: { DataCategory: 'ASSET#TEST' },
            items: [],
            mergeFunction: expect.any(Function),
            extractKey: expect.any(Function)
        })
    })

    //
    // Tests to confirm correct saving of default appearance on Room and Feature Asset records
    //

    it('should save correct default appearance for Rooms and Features', async () => {
        await dbRegister({
            fileName: 'test.wml',
            translateFile: 'test.translate.json',
            scopeMap: {
                Welcome: '12345',
                power: 'ABC',
                togglePower: 'DEF'
            },
            assets: {
                TEST: {
                    tag: 'Asset',
                    key: 'TEST',
                    name: 'Test',
                    zone: 'Library',
                    appearances: [{
                        contextStack: [],
                        contents: [{
                            tag: 'Condition',
                            key: 'Condition-0',
                            index: 0
                        },
                        {
                            tag: 'Room',
                            key: 'Welcome',
                            index: 0
                        },
                        {
                            tag: 'Feature',
                            key: 'clockTower',
                            index: 0
                        }]
                    }]
                },
                ['Condition-0']: {
                    tag: 'Condition',
                    key: 'Condition-0',
                    if: 'false',
                    appearances: [{
                        contextStack: [{ key: 'TEST', tag: 'Asset', index: 0 }],
                        contents: [{
                            tag: 'Room',
                            key: 'Welcome',
                            index: 1
                        },
                        {
                            tag: 'Feature',
                            key: 'clockTower',
                            index: 1
                        }]
                    }]
                },
                Welcome: {
                    tag: 'Room',
                    key: 'Welcome',
                    appearances: [{
                        contextStack: [{ key: 'TEST', tag: 'Asset', index: 0 }],
                        contents: [],
                        render: ['Test render!', { tag: 'Link', to: 'clockTower', text: '(clock tower)'}],
                        name: 'Test'
                    },
                    {
                        contextStack: [{ key: 'TEST', tag: 'Asset', index: 0 }, { key: 'Condition-0', tag: 'Condition', index: 0 }],
                        contents: [],
                        render: ['Should not render'],
                        name: 'Should not'
                    }]
                },
                clockTower: {
                    tag: 'Feature',
                    key: 'clockTower',
                    appearances: [{
                        contextStack: [{ tag: 'Asset', key: 'TEST', index: 0 }],
                        name: 'TestFeature',
                        render: ['Test feature render']
                    },
                    {
                        contextStack: [{ tag: 'Asset', key: 'TEST', index: 0 }],
                    },
                    {
                        contextStack: [{ key: 'TEST', tag: 'Asset', index: 0 }, { key: 'Condition-0', tag: 'Condition', index: 0 }],
                        render: ['Should not render'],
                        name: 'Should not'
                    }]
                }
            }
        })
        expect(assetDB.putItem).toHaveBeenCalledWith({
            AssetId: 'ASSET#TEST',
            DataCategory: 'Meta::Asset',
            fileName: 'test.wml',
            translateFile: 'test.translate.json',
            name: 'Test',
            zone: 'Library'
        })
        expect(mergeIntoDataRange).toHaveBeenCalledWith({
            table: 'assets',
            search: { DataCategory: 'ASSET#TEST' },
            items: [{
                tag: 'Room',
                key: 'Welcome',
                defaultAppearances: [{
                    render: [
                        'Test render!',
                        {
                            tag: 'Link',
                            to: 'clockTower',
                            text: '(clock tower)',
                            targetTag: 'Feature'
                        }
                    ],
                    name: 'Test',
                    contents: []
                }]
            },
            {
                tag: 'Feature',
                key: 'clockTower',
                defaultAppearances: [{
                    name: 'TestFeature',
                    render: ['Test feature render']
                }]
            }],
            mergeFunction: expect.any(Function),
            extractKey: expect.any(Function)
        })
    })
})