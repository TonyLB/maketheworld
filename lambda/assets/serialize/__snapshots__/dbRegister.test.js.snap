// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dbRegister should put a single element for a Character file 1`] = `
Object {
  "AssetId": "CHARACTER#12345",
  "DataCategory": "Meta::Character",
  "Name": "Tess",
  "address": Object {
    "fileName": "test",
    "zone": "Library",
  },
  "fileURL": "testIcon.png",
  "scopedId": "TESS",
}
`;

exports[`dbRegister should save correct default appearance for Rooms and Features 1`] = `
Object {
  "AssetId": "ASSET#TEST",
  "DataCategory": "Meta::Asset",
  "Story": undefined,
  "defaultExits": Array [],
  "defaultNames": Object {
    "Welcome": Object {
      "name": "Test",
      "tag": "Room",
    },
    "clockTower": Object {
      "name": "TestFeature",
      "tag": "Feature",
    },
  },
  "fileName": "test",
  "importTree": Array [],
  "instance": undefined,
  "player": undefined,
  "subFolder": undefined,
  "zone": "Library",
}
`;

exports[`dbRegister should save correct default appearance for Rooms and Features 2`] = `
Object {
  "extractKey": [Function],
  "items": Array [
    Object {
      "defaultAppearances": Array [
        Object {
          "exits": Array [],
          "name": "Test",
          "render": Array [
            "Test render!",
            Object {
              "tag": "Link",
              "targetTag": "Feature",
              "text": "(clock tower)",
              "to": "clockTower",
            },
          ],
        },
      ],
      "key": "Welcome",
      "tag": "Room",
    },
    Object {
      "defaultAppearances": Array [
        Object {
          "name": "TestFeature",
          "render": Array [
            "Test feature render",
          ],
        },
      ],
      "key": "clockTower",
      "tag": "Feature",
    },
  ],
  "mergeFunction": [Function],
  "search": Object {
    "DataCategory": "ASSET#TEST",
  },
  "table": "assets",
}
`;

exports[`dbRegister should save exits in default appearance for Rooms 1`] = `
Object {
  "AssetId": "ASSET#TEST",
  "DataCategory": "Meta::Asset",
  "Story": undefined,
  "defaultExits": Array [
    Object {
      "from": "Entry",
      "name": "welcome",
      "to": "Welcome",
    },
  ],
  "defaultNames": Object {
    "Entry": Object {
      "name": "Entry",
      "tag": "Room",
    },
    "Welcome": Object {
      "name": "Test",
      "tag": "Room",
    },
  },
  "fileName": "test",
  "importTree": Array [],
  "instance": undefined,
  "player": undefined,
  "subFolder": undefined,
  "zone": "Library",
}
`;

exports[`dbRegister should save exits in default appearance for Rooms 2`] = `
Object {
  "extractKey": [Function],
  "items": Array [
    Object {
      "defaultAppearances": Array [
        Object {
          "exits": Array [],
          "name": "Test",
          "render": Array [
            "Test render!",
          ],
        },
      ],
      "key": "Welcome",
      "tag": "Room",
    },
    Object {
      "defaultAppearances": Array [
        Object {
          "exits": Array [
            Object {
              "name": "welcome",
              "to": "Welcome",
            },
          ],
          "name": "Entry",
          "render": Array [
            "Entry render!",
          ],
        },
      ],
      "key": "Entry",
      "tag": "Room",
    },
  ],
  "mergeFunction": [Function],
  "search": Object {
    "DataCategory": "ASSET#TEST",
  },
  "table": "assets",
}
`;

exports[`dbRegister should save meta only for instanced Story type 1`] = `
Object {
  "AssetId": "ASSET#TEST",
  "DataCategory": "Meta::Asset",
  "Story": true,
  "defaultExits": Array [],
  "defaultNames": Object {},
  "fileName": "test",
  "importTree": Array [],
  "instance": true,
  "player": undefined,
  "subFolder": undefined,
  "zone": "Library",
}
`;

exports[`dbRegister should save meta only for instanced Story type 2`] = `
Object {
  "extractKey": [Function],
  "items": Array [],
  "mergeFunction": [Function],
  "search": Object {
    "DataCategory": "ASSET#TEST",
  },
  "table": "assets",
}
`;

exports[`dbRegister should save meta, rooms for Asset type 1`] = `
Object {
  "AssetId": "ASSET#TEST",
  "DataCategory": "Meta::Asset",
  "Story": undefined,
  "defaultExits": Array [],
  "defaultNames": Object {
    "Welcome": Object {
      "name": "Welcome",
      "tag": "Room",
    },
  },
  "fileName": "test",
  "importTree": Array [],
  "instance": undefined,
  "player": undefined,
  "subFolder": undefined,
  "zone": "Library",
}
`;

exports[`dbRegister should save meta, rooms for Asset type 2`] = `
Object {
  "extractKey": [Function],
  "items": Array [
    Object {
      "defaultAppearances": Array [
        Object {
          "rooms": Array [
            Object {
              "name": "Welcome",
              "x": 0,
              "y": 100,
            },
          ],
        },
      ],
      "key": "Village",
      "tag": "Map",
    },
    Object {
      "defaultAppearances": Array [
        Object {
          "exits": Array [],
          "name": "Welcome",
          "render": Array [],
        },
      ],
      "key": "Welcome",
      "tag": "Room",
    },
    Object {
      "defaultAppearances": Array [],
      "key": "clockTower",
      "tag": "Feature",
    },
  ],
  "mergeFunction": [Function],
  "search": Object {
    "DataCategory": "ASSET#TEST",
  },
  "table": "assets",
}
`;

exports[`dbRegister should save meta, rooms for Story type 1`] = `
Object {
  "AssetId": "ASSET#TEST",
  "DataCategory": "Meta::Asset",
  "Story": true,
  "defaultExits": Array [],
  "defaultNames": Object {},
  "fileName": "test",
  "importTree": Array [],
  "instance": undefined,
  "player": undefined,
  "subFolder": undefined,
  "zone": "Library",
}
`;

exports[`dbRegister should save meta, rooms for Story type 2`] = `
Object {
  "extractKey": [Function],
  "items": Array [
    Object {
      "defaultAppearances": Array [],
      "key": "Welcome",
      "tag": "Room",
    },
  ],
  "mergeFunction": [Function],
  "search": Object {
    "DataCategory": "ASSET#TEST",
  },
  "table": "assets",
}
`;
